File: bemkl_supervised_multilabel_classification_variational_train.m
% Mehmet Gonen (mehmet.gonen@gmail.com)

function state = bemkl_supervised_multilabel_classification_variational_train(Km, Y, parameters)
    rand('state', parameters.seed); %#ok<RAND>
    randn('state', parameters.seed); %#ok<RAND>

    D = size(Km, 1);
    N = size(Km, 2);
    P = size(Km, 3);
    L = size(Y, 1);
    sigma_g = parameters.sigma_g;

    log2pi = log(2 * pi);

    Lambda.alpha = (parameters.alpha_lambda + 0.5) * ones(D, L);
    Lambda.beta = parameters.beta_lambda * ones(D, L);
    A.mu = randn(D, L);
    A.sigma = repmat(eye(D, D), [1, 1, L]);
    G.mu = (abs(randn(P, N, L)) + parameters.margin);
    for m = 1:P
        G.mu(m, :, :) = reshape(G.mu(m, :, :), N, L) .* sign(Y');
    end
    G.sigma = eye(P, P);
    gamma.alpha = (parameters.alpha_gamma + 0.5) * ones(L, 1);
    gamma.beta = parameters.beta_gamma * ones(L, 1);
    omega.alpha = (parameters.alpha_omega + 0.5) * ones(P, 1);
    omega.beta = parameters.beta_omega * ones(P, 1);
    be.mu = [zeros(L, 1); ones(P, 1)];
    be.sigma = eye(L + P, L + P);
    F.mu = (abs(randn(L, N)) + parameters.margin) .* sign(Y);
    F.sigma = ones(L, N);
    
    KmKm = zeros(D, D);
    for m = 1:P
        KmKm = KmKm + Km(:, :, m) * Km(:, :, m)';
    end
    Km = reshape(Km, [D, N * P]);

    lower = -1e40 * ones(L, N);
    lower(Y > 0) = +parameters.margin;
    upper = +1e40 * ones(L, N);
    upper(Y < 0) = -parameters.margin;

    if parameters.progress == 1
        bounds = zeros(parameters.iteration, 1);
    end

    atimesaT.mu = zeros(D, D, L);
    for o = 1:L
        atimesaT.mu(:, :, o) = A.mu(:, o) * A.mu(:, o)' + A.sigma(:, :, o);
    end
    GtimesGT.mu = zeros(P, P, L);
    for o = 1:L
        GtimesGT.mu(:, :, o) = G.mu(:, :, o) * G.mu(:, :, o)' + N * G.sigma;
    end
    btimesbT.mu = be.mu(1:L) * be.mu(1:L)' + be.sigma(1:L, 1:L);
    etimeseT.mu = be.mu(L + 1:L + P) * be.mu(L + 1:L + P)' + be.sigma(L + 1:L + P, L + 1:L + P);
    etimesb.mu = zeros(P, L);
    for o = 1:L
        etimesb.mu(:, o) = be.mu(L + 1:L + P) * be.mu(o) + be.sigma(L + 1:L + P, o);
    end
    KmtimesGT.mu = zeros(D, L);
    for o = 1:L
        KmtimesGT.mu(:, o) = Km * reshape(G.mu(:, :, o)', N * P, 1);
    end
    for iter = 1:parameters.iteration
        if mod(iter, 1) == 0
            fprintf(1, '.');
        end
        if mod(iter, 10) == 0
            fprintf(1, ' %5d\n', iter);
        end

        %%%% update Lambda
        for o = 1:L
            Lambda.beta(:, o) = 1 ./ (1 / parameters.beta_lambda + 0.5 * diag(atimesaT.mu(:, :, o)));
        end
        %%%% update A
        for o = 1:L
            A.sigma(:, :, o) = (diag(Lambda.alpha(:, o) .* Lambda.beta(:, o)) + KmKm / sigma_g^2) \ eye(D, D);
            A.mu(:, o) = A.sigma(:, :, o) * KmtimesGT.mu(:, o) / sigma_g^2;
            atimesaT.mu(:, :, o) = A.mu(:, o) * A.mu(:, o)' + A.sigma(:, :, o);
        end
        %%%% update G
        G.sigma = (eye(P, P) / sigma_g^2 + etimeseT.mu) \ eye(P, P);
        for o = 1:L
            G.mu(:, :, o) = G.sigma * (reshape(A.mu(:, o)' * Km, [N, P])' / sigma_g^2 + be.mu(L + 1:L + P) * F.mu(o, :) - repmat(etimesb.mu(:, o), 1, N));
            GtimesGT.mu(:, :, o) = G.mu(:, :, o) * G.mu(:, :, o)' + N * G.sigma;
            KmtimesGT.mu(:, o) = Km * reshape(G.mu(:, :, o)', N * P, 1);
        end
        %%%% update gamma
        gamma.beta = 1 ./ (1 / parameters.beta_gamma + 0.5 * diag(btimesbT.mu));
        %%%% update omega
        omega.beta = 1 ./ (1 / parameters.beta_omega + 0.5 * diag(etimeseT.mu));
        %%%% update b and e
        be.sigma = [diag(gamma.alpha .* gamma.beta) + N * eye(L, L), squeeze(sum(G.mu, 2))'; ...
                         squeeze(sum(G.mu, 2)), diag(omega.alpha .* omega.beta)];
        for o = 1:L
            be.sigma(L + 1:L + P, L + 1:L + P) = be.sigma(L + 1:L + P, L + 1:L + P) + GtimesGT.mu(:, :, o);
        end
        be.sigma = be.sigma \ eye(L + P, L + P);
        be.mu = zeros(L + P, 1);
        be.mu(1:L) = sum(F.mu, 2);
        for o = 1:L
            be.mu(L + 1:L + P) = be.mu(L + 1:L + P) + G.mu(:, :, o) * F.mu(o, :)';
        end
        be.mu = be.sigma * be.mu;
        btimesbT.mu = be.mu(1:L) * be.mu(1:L)' + be.sigma(1:L, 1:L);
        etimeseT.mu = be.mu(L + 1:L + P) * be.mu(L + 1:L + P)' + be.sigma(L + 1:L + P, L + 1:L + P);
        for o = 1:L
            etimesb.mu(:, o) = be.mu(L + 1:L + P) * be.mu(o) + be.sigma(L + 1:L + P, o);
        end
        %%%% update F
        output = zeros(L, N);
        for o = 1:L
            output(o, :) = [ones(1, N); G.mu(:, :, o)]' * be.mu([o, L + 1:L + P]);
        end
        alpha_norm = lower - output;
        beta_norm = upper - output;
        normalization = normcdf(beta_norm) - normcdf(alpha_norm);
        normalization(normalization == 0) = 1;
        F.mu = output + (normpdf(alpha_norm) - normpdf(beta_norm)) ./ normalization;
        F.sigma = 1 + (alpha_norm .* normpdf(alpha_norm) - beta_norm .* normpdf(beta_norm)) ./ normalization - (normpdf(alpha_norm) - normpdf(beta_norm)).^2 ./ normalization.^2;

        if parameters.progress == 1
            lb = 0;

            %%%% p(Lambda)
            lb = lb + sum(sum((parameters.alpha_lambda - 1) * (psi(Lambda.alpha) + log(Lambda.beta)) ...
                              - Lambda.alpha .* Lambda.beta / parameters.beta_lambda ...
                              - gammaln(parameters.alpha_lambda) ...
                              - parameters.alpha_lambda * log(parameters.beta_lambda)));
            %%%% p(A | Lambda)
            for o = 1:L
                lb = lb - 0.5 * sum(Lambda.alpha(:, o) .* Lambda.beta(:, o) .* diag(atimesaT.mu(:, :, o))) ...
                        - 0.5 * (D * log2pi - sum(log(Lambda.alpha(:, o) .* Lambda.beta(:, o))));
            end
            %%%% p(G | A, Km)
            for o = 1:L
                lb = lb - 0.5 * sum(diag(GtimesGT.mu(:, :, o))) ...
                        + A.mu(:, o)' * KmtimesGT.mu(:, o) ...
                        - 0.5 * sum(sum(KmKm .* atimesaT.mu(:, :, o))) ...
                        - 0.5 * N * P * (log2pi + 2 * log(sigma_g));
            end
            %%%% p(gamma)
            lb = lb + sum((parameters.alpha_gamma - 1) * (psi(gamma.alpha) + log(gamma.beta)) ...
                          - gamma.alpha .* gamma.beta / parameters.beta_gamma ...
                          - gammaln(parameters.alpha_gamma) ...
                          - parameters.alpha_gamma * log(parameters.beta_gamma));
            %%%% p(b | gamma)
            lb = lb - 0.5 * sum(gamma.alpha .* gamma.beta .* diag(btimesbT.mu)) ...
                    - 0.5 * (L * log2pi - sum(log(gamma.alpha .* gamma.beta)));
            %%%% p(omega)
            lb = lb + sum((parameters.alpha_omega - 1) * (psi(omega.alpha) + log(omega.beta)) ...
                          - omega.alpha .* omega.beta / parameters.beta_omega ...
                          - gammaln(parameters.alpha_omega) ...
                          - parameters.alpha_omega * log(parameters.beta_omega));
            %%%% p(e | omega)
            lb = lb - 0.5 * sum(omega.alpha .* omega.beta .* diag(etimeseT.mu)) ...
                    - 0.5 * (P * log2pi - sum(log(omega.alpha .* omega.beta)));
            %%%% p(F | b, e, G)
            for o = 1:L
                lb = lb - 0.5 * (F.mu(o, :) * F.mu(o, :)' + sum(F.sigma(o, :))) ...
                        + F.mu(o, :) * (G.mu(:, :, o)' * be.mu(L + 1:L + P)) ...
                        + sum(be.mu(o) * F.mu(o, :)) ...
                        - 0.5 * sum(sum(etimeseT.mu .* GtimesGT.mu(:, :, o))) ...
                        - sum(G.mu(:, :, o)' * etimesb.mu(:, o)) ...
                        - 0.5 * N * btimesbT.mu(o, o) ...
                        - 0.5 * N * log2pi;
            end

            %%%% q(Lambda)
            lb = lb + sum(sum(Lambda.alpha + log(Lambda.beta) + gammaln(Lambda.alpha) + (1 - Lambda.alpha) .* psi(Lambda.alpha)));
            %%%% q(A)
            for o = 1:L
                lb = lb + 0.5 * (D * (log2pi + 1) + logdet(A.sigma(:, :, o)));
            end
            %%%% q(G)
            lb = lb + 0.5 * L * (N * (P * (log2pi + 1) + logdet(G.sigma)));
            %%%% q(gamma)
            lb = lb + sum(gamma.alpha + log(gamma.beta) + gammaln(gamma.alpha) + (1 - gamma.alpha) .* psi(gamma.alpha));
            %%%% q(omega)
            lb = lb + sum(omega.alpha + log(omega.beta) + gammaln(omega.alpha) + (1 - omega.alpha) .* psi(omega.alpha));
            %%%% q(b, e)
            lb = lb + 0.5 * ((L + P) * (log2pi + 1) + logdet(be.sigma));
            %%%% q(F)
            lb = lb + 0.5 * sum(sum(log2pi + F.sigma)) + sum(sum(log(normalization)));

            bounds(iter) = lb;
        end
    end

    state.Lambda = Lambda;
    state.A = A;
    state.gamma = gamma;
    state.omega = omega;
    state.be = be;
    if parameters.progress == 1
        state.bounds = bounds;
    end
    state.parameters = parameters;
end

function ld = logdet(Sigma)
    U = chol(Sigma);
    ld = 2 * sum(log(diag(U)));
end

File: bemkl_supervised_multilabel_classification_variational_train.R
# Mehmet Gonen (mehmet.gonen@gmail.com)

logdet <- function(Sigma) {
    2 * sum(log(diag(chol(Sigma))))
}

bemkl_supervised_multilabel_classification_variational_train <- function(Km, Y, parameters) {
  set.seed(parameters$seed)

  D <- dim(Km)[1]
  N <- dim(Km)[2]
  P <- dim(Km)[3]
  L <- dim(Y)[1]
  sigma_g <- parameters$sigma_g

  log2pi <- log(2 * pi)

  Lambda <- list(alpha = matrix(parameters$alpha_lambda + 0.5, D, L), beta = matrix(parameters$beta_lambda, D, L))
  A <- list(mu = matrix(rnorm(D * L), D, L), sigma = array(diag(1, D, D), c(D, D, L)))
  G <- list(mu = (abs(array(rnorm(P * N * L), c(P, N, L))) + parameters$margin), sigma = diag(1, P, P))
  for (m in 1:P) {
    G$mu[m,,] <- G$mu[m,,] * sign(t(Y))
  }
  gamma <- list(alpha = matrix(parameters$alpha_gamma + 0.5, L, 1), beta = matrix(parameters$beta_gamma, L, 1))
  omega <- list(alpha = matrix(parameters$alpha_omega + 0.5, P, 1), beta = matrix(parameters$beta_omega, P, 1))
  be <- list(mu = rbind(matrix(0, L, 1), matrix(1, P, 1)), sigma = diag(1, L + P, L + P))
  F <- list(mu = (abs(matrix(rnorm(L * N), L, N)) + parameters$margin) * sign(Y), sigma = matrix(1, L, N))

  KmKm <- matrix(0, D, D)
  for (m in 1:P) {
    KmKm <- KmKm + tcrossprod(Km[,,m], Km[,,m])
  }
  Km <- matrix(Km, D, N * P)

  lower <- matrix(-1e40, L, N)
  lower[which(Y > 0)] <- +parameters$margin
  upper <- matrix(+1e40, L, N)
  upper[which(Y < 0)] <- -parameters$margin
  
  if (parameters$progress == 1) {
    bounds <- matrix(0, parameters$iteration, 1)
  }

  atimesaT.mu <- array(0, c(D, D, L))
  for (o in 1:L) {
    atimesaT.mu[,,o] <- tcrossprod(A$mu[,o], A$mu[,o]) + A$sigma[,,o]
  }
  GtimesGT.mu <- array(0, c(P, P, L))
  for (o in 1:L) {
    GtimesGT.mu[,,o] <- tcrossprod(G$mu[,,o], G$mu[,,o]) + N * G$sigma
  }
  btimesbT.mu <- tcrossprod(be$mu[1:L], be$mu[1:L]) + be$sigma[1:L, 1:L]
  etimeseT.mu <- tcrossprod(be$mu[(L + 1):(L + P)], be$mu[(L + 1):(L + P)]) + be$sigma[(L + 1):(L + P), (L + 1):(L + P)]
  etimesb.mu <- matrix(0, P, L)
  for (o in 1:L) {
    etimesb.mu[,o] <- be$mu[(L + 1):(L + P)] * be$mu[o] + be$sigma[(L + 1):(L + P), o]
  }
  KmtimesGT.mu <- matrix(0, D, L)
  for (o in 1:L) {
    KmtimesGT.mu[,o] <- Km %*% matrix(t(G$mu[,,o]), N * P, 1)
  }
  for (iter in 1:parameters$iteration) {
    # update Lambda
    for (o in 1:L) {
      Lambda$beta[,o] <- 1 / (1 / parameters$beta_lambda + 0.5 * diag(atimesaT.mu[,,o]))
    }
    # update A
    for (o in 1:L) {
      A$sigma[,,o] <- chol2inv(chol(diag(as.vector(Lambda$alpha[,o] * Lambda$beta[,o]), D, D) + KmKm / sigma_g^2))
      A$mu[,o] <- A$sigma[,,o] %*% KmtimesGT.mu[,o] / sigma_g^2
      atimesaT.mu[,,o] <- tcrossprod(A$mu[,o], A$mu[,o]) + A$sigma[,,o]
    }
    # update G
    G$sigma <- chol2inv(chol(diag(1, P, P) / sigma_g^2 + etimeseT.mu))
    for (o in 1:L) {
      G$mu[,,o] <- G$sigma %*% (t(matrix(crossprod(A$mu[,o], Km), N, P)) / sigma_g^2 + tcrossprod(be$mu[(L + 1):(L + P)], F$mu[o,]) - matrix(etimesb.mu[,o], P, N, byrow = FALSE))
      GtimesGT.mu[,,o] <- tcrossprod(G$mu[,,o], G$mu[,,o]) + N * G$sigma
      KmtimesGT.mu[,o] <- Km %*% matrix(t(G$mu[,,o]), N * P, 1)
    }
    # update gamma
    gamma$beta <- 1 / (1 / parameters$beta_gamma + 0.5 * diag(btimesbT.mu))
    # update omega
    omega$beta <- 1 / (1 / parameters$beta_omega + 0.5 * diag(etimeseT.mu))
    # update b and e
    be$sigma <- rbind(cbind(diag(as.vector(gamma$alpha * gamma$beta), L, L) + N * diag(1, L, L), t(apply(G$mu, c(1, 3), sum))), cbind(apply(G$mu, c(1, 3), sum), diag(as.vector(omega$alpha * omega$beta), P, P)))
    for (o in 1:L) {
      be$sigma[(L + 1):(L + P), (L + 1):(L + P)] <- be$sigma[(L + 1):(L + P), (L + 1):(L + P)] + GtimesGT.mu[,,o]
    }
    be$sigma <- chol2inv(chol(be$sigma))
    be$mu <- matrix(0, L + P, 1)
    be$mu[1:L] <- rowSums(F$mu)
    for (o in 1:L) {
      be$mu[(L + 1):(L + P)] <- be$mu[(L + 1):(L + P)] + G$mu[,,o] %*% F$mu[o,]
    }
    be$mu <- be$sigma %*% be$mu
    btimesbT.mu <- tcrossprod(be$mu[1:L], be$mu[1:L]) + be$sigma[1:L, 1:L]
    etimeseT.mu <- tcrossprod(be$mu[(L + 1):(L + P)], be$mu[(L + 1):(L + P)]) + be$sigma[(L + 1):(L + P), (L + 1):(L + P)]
    for (o in 1:L) {
        etimesb.mu[,o] <- be$mu[(L + 1):(L + P)] * be$mu[o] + be$sigma[(L + 1):(L + P), o]
    }
    # update F
    output <- matrix(0, L, N)
    for (o in 1:L) {
      output[o,] <- crossprod(rbind(matrix(1, 1, N), G$mu[,,o]), be$mu[c(o, (L + 1):(L + P))])
    }
    alpha_norm <- lower - output
    beta_norm <- upper - output
    normalization <- pnorm(beta_norm) - pnorm(alpha_norm)
    normalization[which(normalization == 0)] <- 1
    F$mu <- output + (dnorm(alpha_norm) - dnorm(beta_norm)) / normalization
    F$sigma <- 1 + (alpha_norm * dnorm(alpha_norm) - beta_norm * dnorm(beta_norm)) / normalization - (dnorm(alpha_norm) - dnorm(beta_norm))^2 / normalization^2

    if (parameters$progress == 1) {
      lb <- 0

      # p(Lambda)
      lb <- lb + sum((parameters$alpha_lambda - 1) * (digamma(Lambda$alpha) + log(Lambda$beta)) - Lambda$alpha * Lambda$beta / parameters$beta_lambda - lgamma(parameters$alpha_lambda) - parameters$alpha_lambda * log(parameters$beta_lambda))
      # p(A | Lambda)
      for (o in 1:L) {
        lb <- lb - 0.5 * sum(as.vector(Lambda$alpha[,o] * Lambda$beta[,o]) * diag(atimesaT.mu[,,o])) - 0.5 * (D * log2pi - sum(digamma(Lambda$alpha[,o]) + log(Lambda$beta[,o])))
      }
      # p(G | A, Km)
      for (o in 1:L) {
        lb <- lb - 0.5 * sigma_g^-2 * sum(diag(GtimesGT.mu[,,o])) + sigma_g^-2 * crossprod(A$mu[,o], KmtimesGT.mu[,o]) - 0.5 * sigma_g^-2 * sum(KmKm * atimesaT.mu[,,o]) - 0.5 * N * P * (log2pi + 2 * log(sigma_g))
      }
      # p(gamma)
      lb <- lb + sum((parameters$alpha_gamma - 1) * (digamma(gamma$alpha) + log(gamma$beta)) - gamma$alpha * gamma$beta / parameters$beta_gamma - lgamma(parameters$alpha_gamma) - parameters$alpha_gamma * log(parameters$beta_gamma))
      # p(b | gamma)
      lb <- lb - 0.5 * sum(as.vector(gamma$alpha * gamma$beta) * diag(btimesbT.mu)) - 0.5 * (L * log2pi - sum(digamma(gamma$alpha) + log(gamma$beta)))
      # p(omega)
      lb <- lb + sum((parameters$alpha_omega - 1) * (digamma(omega$alpha) + log(omega$beta)) - omega$alpha * omega$beta / parameters$beta_omega - lgamma(parameters$alpha_omega) - parameters$alpha_omega * log(parameters$beta_omega))
      # p(e | omega)
      lb <- lb - 0.5 * sum(as.vector(omega$alpha * omega$beta) * diag(etimeseT.mu)) - 0.5 * (P * log2pi - sum(digamma(omega$alpha) + log(omega$beta)))
      # p(F | b, e, G)
      for (o in 1:L) {
        lb <- lb - 0.5 * (crossprod(F$mu[o,], F$mu[o,]) + sum(F$sigma[o,])) + crossprod(F$mu[o,], crossprod(G$mu[,,o], be$mu[(L + 1):(L + P)])) + sum(be$mu[o] * F$mu[o,]) - 0.5 * sum(etimeseT.mu * GtimesGT.mu[,,o]) - sum(crossprod(G$mu[,,o], etimesb.mu[,o])) - 0.5 * N * btimesbT.mu[o,o] - 0.5 * N * log2pi
      }

      # q(Lambda)
      lb <- lb + sum(Lambda$alpha + log(Lambda$beta) + lgamma(Lambda$alpha) + (1 - Lambda$alpha) * digamma(Lambda$alpha))
      # q(A)
      for (o in 1:L) {
        lb <- lb + 0.5 * (D * (log2pi + 1) + logdet(A$sigma[,,o]))
      }
      # q(G)
      lb <- lb + 0.5 * L * (N * (P * (log2pi + 1) + logdet(G$sigma)))
      # q(gamma)
      lb <- lb + sum(gamma$alpha + log(gamma$beta) + lgamma(gamma$alpha) + (1 - gamma$alpha) * digamma(gamma$alpha))
      # q(omega)
      lb <- lb + sum(omega$alpha + log(omega$beta) + lgamma(omega$alpha) + (1 - omega$alpha) * digamma(omega$alpha))
      # q(b, e)
      lb <- lb + 0.5 * ((L + P) * (log2pi + 1) + logdet(be$sigma))
      # q(F)
      lb <- lb + 0.5 * sum(log2pi + F$sigma) + sum(log(normalization))

      bounds[iter] <- lb
    }
  }
  
  if (parameters$progress == 1) {
    state <- list(Lambda = Lambda, A = A, gamma = gamma, omega = omega, be = be, bounds = bounds, parameters = parameters)
  }
  else {
    state <- list(Lambda = Lambda, A = A, gamma = gamma, omega = omega, be = be, parameters = parameters)
  }
}